File: default/typescript/lib/pipeline-stack.ts
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { CodePipeline, CodePipelineSource, ShellStep, ManualApprovalStep } from 'aws-cdk-lib/pipelines';
import { AppStage } from './app-stage';

/**
 * Default CDK pipeline stack with GitHub source and multi-stage deployment.
 * Includes automated testing and manual approval for production deployment.
 */
export class PipelineStack extends cdk.Stack {
  /**
   * Creates a new PipelineStack instance.
   * @param scope - The scope in which to define this construct
   * @param id - The scoped construct ID
   * @param props - Stack properties
   */
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    /** Self-mutating CDK pipeline with GitHub source */
    const pipeline = new CodePipeline(this, 'Pipeline', {
      pipelineName: 'DefaultPipeline',
      synth: new ShellStep('Synth', {
        input: CodePipelineSource.gitHub('aws-samples/aws-cdk-examples', 'main'),
        commands: [
          'npm ci',
          'npm run build',
          'npx cdk synth'
        ]
      }),
      dockerEnabledForSynth: true,
      dockerEnabledForSelfMutation: true,
    });

    /** Test stage with automated unit testing */
    const testStage = pipeline.addStage(new AppStage(this, 'Test', {
      env: { 
        account: process.env.CDK_DEFAULT_ACCOUNT, 
        region: process.env.CDK_DEFAULT_REGION 
      }
    }));

    /** Add unit tests before test deployment */
    testStage.addPre(new ShellStep('UnitTest', {
      commands: [
        'npm ci',
        'npm test'
      ]
    }));

    /** Production stage with manual approval gate */
    const prodStage = pipeline.addStage(new AppStage(this, 'Prod', {
      env: { 
        account: process.env.CDK_DEFAULT_ACCOUNT, 
        region: process.env.CDK_DEFAULT_REGION 
      }
    }));

    /** Require manual approval before production deployment */
    prodStage.addPre(new ManualApprovalStep('PromoteToProd'));
  }
}

File: default/csharp/src/PipelineStack.cs
using Amazon.CDK;
using Amazon.CDK.AWS.CodeBuild;
using Amazon.CDK.AWS.IAM;
using Amazon.CDK.Pipelines;
using Constructs;

/// <summary>
/// Pipeline stack with multi-stage deployment.
/// </summary>
namespace Default
{
    public class PipelineStack : Stack
    {
        internal PipelineStack(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
        {
            var pipeline = new CodePipeline(this, "Pipeline", new CodePipelineProps
            {
                PipelineName = "defaultPipeline",
                Synth = new ShellStep("Synth", new ShellStepProps
                {
                    Input = CodePipelineSource.GitHub("OWNER/REPO", "main"),
                    Commands = new[] { "dotnet build src", "npx cdk synth" }
                }),
                SynthCodeBuildDefaults = new CodeBuildOptions
                {
                    BuildEnvironment = new BuildEnvironment
                    {
                        BuildImage = LinuxBuildImage.STANDARD_7_0
                    },
                    RolePolicy = new[]
                    {
                        new PolicyStatement(new PolicyStatementProps
                        {
                            Effect = Effect.ALLOW,
                            Actions = new[]
                            {
                                "ec2:DescribeAvailabilityZones",
                                "ec2:DescribeVpcs",
                                "ec2:DescribeSubnets",
                                "ec2:DescribeRouteTables",
                                "ec2:DescribeSecurityGroups",
                                "ssm:GetParameter",
                                "ssm:GetParameters"
                            },
                            Resources = new[] { "*" }
                        })
                    }
                },
                SelfMutationCodeBuildDefaults = new CodeBuildOptions
                {
                    BuildEnvironment = new BuildEnvironment
                    {
                        BuildImage = LinuxBuildImage.STANDARD_7_0
                    }
                }
            });

            var testStage = pipeline.AddStage(new AppStage(this, "Test", new StageProps
            {
                Env = new Amazon.CDK.Environment
                {
                    Account = System.Environment.GetEnvironmentVariable("CDK_DEFAULT_ACCOUNT"),
                    Region = System.Environment.GetEnvironmentVariable("CDK_DEFAULT_REGION")
                }
            }));

            testStage.AddPre(new ShellStep("UnitTest", new ShellStepProps
            {
                Commands = new[] { "dotnet test src" }
            }));

            var prodStage = pipeline.AddStage(new AppStage(this, "Prod", new StageProps
            {
                Env = new Amazon.CDK.Environment
                {
                    Account = System.Environment.GetEnvironmentVariable("CDK_DEFAULT_ACCOUNT"),
                    Region = System.Environment.GetEnvironmentVariable("CDK_DEFAULT_REGION")
                }
            }));

            prodStage.AddPre(new ManualApprovalStep("PromoteToProd"));
        }
    }
}
